pkgs/swiftgen/swift2objc/lib/src/parser/parser.dart

```diff
@@ -0,0 +1,16 @@
+import '_core/utils.dart';
+import 'parsers/parse_declarations_map.dart';
+import 'parsers/wire_up_relations.dart';
+import 'parsers/parse_symbols_map.dart';
+
+DeclarationsMap parseAst(String symbolgraphJsonPath) {
+  final symbolgraph = readJsonFile(symbolgraphJsonPath);
+
+  final parsedSymbols = parseSymbolsMap(symbolgraph);
+
+  final declarations = parseDeclarationsMap(parsedSymbols);
+
+  wireUpRelations(declarations, symbolgraph);
```

Comment:
The structure of this `parseAst` function is good overall. I have an idea to make it even cleaner, but I'm not sure if it'll work, so feel free to keep things how they are if it's a bad idea.

The reason `parseSymbolsMap` is such a clean and simple function is that it doesn't really know anything about the symbols it's looking for. It's just extracting an ID and generating a map from ID to json chunk. Instead, the only function that needs to know the structure of the symbols is `parseDeclarationsMap`.

I'm wondering if it's possible to do the same thing for `wireUpRelations`. Can we move `wireUpRelations` above `parseDeclarationsMap`, and have it just extract the IDs of each relation, and generate a map from ID to json chunks? The return type would look like `Map<String, List<JsonMap>>`. Then we pass this map to `parseDeclarationsMap`, along with the `parsedSymbols`. That way, the only function that needs to understand the relationship semantics is `parseDeclarationsMap`.

For example, `parseClassDeclaration` would also be in charge of adding all the methods to the class. So all the logic that understands the class's json structure, as well as all the logic that understands the relationships between classes and methods, would live in one place.

One problem I can see with this approach is that relationships have both a source and a target ID, and the relationship parsing function wouldn't necessarily know which ID to use as the map key (for example, class methods store the class's ID as the target, not the source, but the `parseClassDeclaration` function would want to look up all the relationships using the class's ID). So I think you'd have to use both IDs as keys:

```Dart
final parsedRelations = <String, List<JsonMap>>{};
for (final relation in relations) {
  parsedRelations[relation["source"]] = relation;
  parsedRelations[relation["target"]] = relation;
}
```

WDYT?

Reply:
I gave it a shot. It's working pretty well I would say.

I created a `ParsedSymbolgraph` model that contains symbols and relations maps (basically the main 2 pieces of data in the json). Now the parsing flow is composed of two main steps:

1. Parse symbolgraph json:
    a. Parse symbols
    b. Parse relations

2. Generate the AST

As you said the nice thing about this is all `json => ast` knowledge is only needed in the second step. The first step is kinda just exploring the symbolgraph and parsing it into an easier data model to work with in the second step.